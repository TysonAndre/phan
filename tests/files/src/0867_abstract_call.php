<?php declare(strict_types = 1);

namespace NS867;

abstract class Base {
    public static function callAbstract() : void {
        self::bar();  // should warn
        static::bar();  // should warn
    }

    public function callInstance() : void {
        self::bar();  // should warn
        static::bar();  // should not warn because $this is not an abstract class.
    }

    abstract static function bar() : void;
}

class A extends Base {
    public static function bar() : void {
        echo "A::bar() called succesfully\n";
    }
}

A::callAbstract(); // fine
Base::callAbstract(); // leads to fatal error, but not tracked by Phan
Base::bar();  // should warn
A::bar();  // should not warn

interface I {
    public static function testMethod();
}
I::testMethod();

trait T {
    public static function generate() {
        $x = static::callAbstract();
        $y = self::callAbstract();
        return [$x, $y];
    }
    public static abstract function callAbstract();
}
abstract class X {
    use T;
}
X::generate();  // this doesn't warn, but Phan already warns about T::generate()
