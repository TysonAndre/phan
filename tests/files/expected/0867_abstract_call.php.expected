%s:7 PhanAbstractStaticMethodCall Potentially calling an abstract static method \NS867\Base::bar() in self::bar()
%s:8 PhanAbstractStaticMethodCallInStatic Potentially calling an abstract static method \NS867\Base::bar() with static:: in static::bar() (the calling static method's class scope may be an abstract class)
%s:12 PhanAbstractStaticMethodCall Potentially calling an abstract static method \NS867\Base::bar() in self::bar()
%s:27 PhanAbstractStaticMethodCall Potentially calling an abstract static method \NS867\Base::bar() in Base::bar()
%s:33 PhanAbstractStaticMethodCall Potentially calling an abstract static method \NS867\I::testMethod() in I::testMethod()
%s:37 PhanAbstractStaticMethodCallInTrait Potentially calling an abstract static method \NS867\T::callAbstract() on a trait in static::callAbstract(), if the caller's method is called on the trait instead of a concrete class using the trait
%s:38 PhanAbstractStaticMethodCallInTrait Potentially calling an abstract static method \NS867\T::callAbstract() on a trait in self::callAbstract(), if the caller's method is called on the trait instead of a concrete class using the trait
